mitrofanova.alina/lab2/host/host.cpp

```diff
@@ -0,0 +1,67 @@
+#include <syslog.h>
+#include <cstdlib>
+#include <unistd.h>
+#include <sys/stat.h>
+#include <signal.h>
+
+#include "ShM.h"
+#include "Semaphore.h"
+#include "wolf.h"
+#include "goat.h"
+
+int main(int argc, char** argv) {
+	openlog("Game 'The wolf and the seven goats'", LOG_CONS | LOG_PID, LOG_USER);
+	syslog(LOG_INFO, "The host is running");
+
+	const char* hostShmName = "host_shm";
+	int hostShm = ShM::Open(hostShmName);
+	sem_t* hostSem = Semaphore::Init(hostShm);
+	
+	const char* clientShmName = "client_shm";
+	int clientShm = ShM::Open(clientShmName);
+	sem_t* clientSem = Semaphore::Init(clientShm);
```

Comment:
Зачем вам разделяемая память для создания семафоров?

Reply:
Разделяемая память - наилучший механизм для межпроцессного взаимодействия. Ну а для синхронизации (атомарности обновлений) процессов используются семафоры. Семафоры должны располагаться на разделяемой памяти. Даже если у нас родственные процессы, все равно надо семафоры располагать на разделяемой памяти.
Я, например, пробовала без разделяемой памяти записать в goat в переменную число, а в wolf вывести значение этой переменной. И вот wolf не видел это число.

Reply:
Нет, не должны. Это глобальные семафоры. Просто откройте их по имени в двух процессах. И почему вы решили, что разделяемая память - наилучший механизм?

Reply:
На самом деле, прочитала в интернете:

> Shared memory offers the highest bandwidth IPC available. 

> Once a shared-memory object is created, processes with access to the object can use pointers to directly read and write into it. This means that access to shared memory is in itself unsynchronized. If a process is updating an area of shared memory, care must be taken to prevent another process from reading or updating the same area. Even in the simple case of a read, the other process may get information that is in flux and inconsistent.

> To solve these problems, shared memory is often used in conjunction with one of the synchronization primitives to make updates atomic between processes. If the granularity of updates is small, then the synchronization primitives themselves will limit the inherently high bandwidth of using shared memory. Shared memory is therefore most efficient when used for updating large amounts of data as a block.
 
> Both semaphores and mutexes are suitable synchronization primitives for use with shared memory. Semaphores were introduced with the POSIX realtime standard for interprocess synchronization. Mutexes were introduced with the POSIX threads standard for thread synchronization.

Reply:
Я не думала о том, что можно открыть семафоры по имени